### ISSUE TITLE
Feature request: configuration resource for naming resources of the same type

### ISSUE THREAD

**Issue by ttzeng** [01org/iot-js-api#26]
In the OCF core spec v1.1.0, section 11.2, though the configuration information resource /oic/con ('oic.wk'con' resource type) is an optional resource, it provides a mechanism for clients to "name" resources to the user friendly meaning (e.g. Bob's thermostat), this will be particular valuable if there are multiple resources of the same type on the network (e.g. light bulbs in a smart home). Right now the APIs will host 'oic.wk.p' and 'oic.wk.d' resources automatically, it would be good to extend the API to 'oic.wk.con' resource as well.

**Comment by zolkis** [comment]
This is currently possible by updating the `/oic/con` resource on the device in interest. But are you asking an explicit JS API for this use case of renaming resources? 

`oic.wk.con` has only one mandatory read-write property, user-defined device name (`"n"`). Optional read-write properties are location (`"loc"`), user-defined alias location name ('"locn"'), currency (`"c"`) and region (`"r"`). 

@gvancuts do we have a developer need for this?

It is also related to the more generic topic of provisioning & configuration, which is a worm can at the moment.

**Comment by gvancuts** [comment]
Hi @zolkis , what we'd like to do is be able to set a user-friendly name from the cloud portal. This is useful when we have multiple devices of the same type in the network, so a user can set that user-friendly name to something more meaningful to her/him (i.e. "Garage temperature sensor", "Zoltan's room temperature sensor", etc.). The idea is that it's a user action to establish this and so would typically be done from an OIC client of some sort (e.g. cloud portal which uses the iot-rest-api-server or an Android app).

From my (basic) understanding of the OIC specs and the the JS bindings, it looks like this would be done by implementing an OcfConfiguration object?

You mentionned that it's possible to update the `/oic/con` resource on the device. How would that be done today?

**Comment by zolkis** [comment]
> You mentioned that it's possible to update the /oic/con resource on the device. How would that be done today?

```javascript
var client = require("ocf").client;
// discover the /oic/con resource on the target device
// or retrieve it directly
client.retrieve({ deviceId: "xxx", resourcePath: "/oic/con" })
.then(function(con) {
    con.properties["n"] = "myLivingRoomTemperatureSensor";
    client.update(con);
}).catch (function(err) {
    // error
});
```
If we define an API, this may look like:
```javascript
var client = require("ocf").client;

client.configure("<deviceUUID>", { name: "myLivingRoomTemperatureSensor" });
}).catch (function(err) {
    // error, probably device not found
});
```
configure() would take a device id (string uuid) and a dictionary like
```javascript
{
  name: "string",
  location: coordinates,  // dictionary
  locationName: "my weekend house",
  currency: "EUR",
  region: "Finland"
}
```

We could place the `configure()` method on the `OcfDevice` object, but that would be bad since then first the device would need to be discovered, whereas it should be possible to configure the device straight away by knowing its UUID.

**Comment by gvancuts** [comment]
Thanks for these details. How would such `/oic/con` resource be implemented when the OCF server is itself written in JS?

**Comment by zolkis** [comment]
It's just another resource that is listed in the `/oic/res` linked list entry and maintained by the _implementation_ of the JS API, like it already has to maintain `/oic/res`, `/oic/d`, `/oic/p`, `/oic/mnt` etc.

In my view, client code doesn't need to explicitly define these meta-resources: by default they should be maintained by the implementation, in the sense of responding to OCF requests with those resource paths. 

Of course, server side JS code can be written for them, for instance the server app can catch requests coming to those resource paths and send responses from JS. The `OcfRequest` is then handled by the app. 

If in turn the app does not define handlers for requests with those resource paths, the implementation SHOULD do it.

@gabrielschulhof please let's check these :).

**Comment by zolkis** [comment]
In general, as a developer I expect the API implementation would take care of as much OCF specific requirements and quirks as possible, so that developers could concentrate on writing code only for supporting their business model related resource properties. 

This simplifies user code, make testing of OCF related stuff early and more standard, so I suggest we do that.

**Comment by ttzeng** [comment]
It looks to me the server scripts using iotivity-node can provide device.device and device.platform objects to config the oic.wk.d and oic.wk.p resources, does it make sense to have similar way to config the oic.wk.con resource?

**Comment by gvancuts** [comment]
> In general, as a developer I expect the API implementation would take care of as much OCF specific requirements and quirks as possible, so that developers could concentrate on writing code only for supporting their business model related resource
> properties.
> This simplifies user code, make testing of OCF related stuff early and more standard, so I suggest we do that.

Completely agree with you here! +1

**Comment by ttzeng** [comment]
In our case, we (i.e. the smart home OCF servers) are developing user code which uses iotivity-node as the API implementation, does that mean we should file the issue / feature request to the iotivity-node project? @poussa referred me to this list.

**Comment by zolkis** [comment]
@ttzeng 
Don't file issue to iotivity-node, since it does not yet have any explicit requirements to encapsulate `/oic/con`.

We could have the configuration API feature in the next API version. In the meantime, use the well-known resource `/oic/con` from applications, as in the examples above.

**Comment by zolkis** [comment]
Please check if the issue is addressed by #30
